hammers/scripts/gpu_lease_stacking.py

```diff
@@ -8,80 +8,145 @@
+
+        for node_id, leases in self.nodes.items():
+            stacked_leases = self.find_stacked_leases(leases)
+            self.leases_to_delete.extend(stacked_leases[LEASES_ALLOWED:])
+
+    def find_stacked_leases(self, leases):
+        """Return list of only the leases stacked on each other."""
+        stacked = []
+
+        for i in range(len(leases)):
+
+            _, start_date, end_date = leases[i]
+
+            if i > 0:
+                last_end_date = leases[i - 1][2]
+            else:
+                last_end_date = datetime.min
+
+            if i < len(leases) - 1:
+                next_start_date = leases[i + 1][1]
+            else:
+                next_start_date = datetime.max
+
+            stacked_previous = (start_date - last_end_date).days < 1
+            stacked_next = (next_start_date - end_date).days < 1
+
+            if stacked_previous or stacked_next:
+                stacked.append(leases[i])
+
+        return stacked
+
+    def send_delete_notification(self, sender):
+        """Send email notifying user of leases deleted."""
+        email_body = _email.STACKED_LEASE_DELETED_EMAIL_BODY
+        html = _email.render_template(
+            email_body,
+            vars={'lease_list': [x[0] for x in self.leases_to_delete]})
+        subject = "Your GPU lease(s) was deleted."
+        _email.send(
+            _email.get_host(),
+            self.email,
+            sender,
+            subject,
+            html.encode("utf8"))
+
+    def print_info(self):
+        """Return dict of info for console output."""
+        return {
+            'user_name': self.name,
+            'user_email': self.email,
+            'leases': self.leases_to_delete
+        }
+
+
+def gpu_stack_reaper(db, auth, sender, describe=False, quiet=False):
     """Delete stacked leases on gpu nodes."""
-    user_gpu_leases = {}
+    gpu_users = {}
 
-    for row in query.gpu_leases(db):
+    for row in query.get_gpu_advanced_reservations(db):
         user_id = row['user_id']
-        node_id = row['node_id']
-
-        if user_id not in user_gpu_leases.keys():
-            user_gpu_leases[user_id] = {}
 
-        if node_id not in user_gpu_leases[user_id].keys():
-            user_gpu_leases[user_id][node_id] = []
+        if user_id not in gpu_users.keys():
+            gpu_users[user_id] = GPUUser(
+                user_id=user_id,
+                name=row['user_name'],
+                email=json.loads(row['user_extra'])['email'])
 
-        user_gpu_leases[user_id][node_id].append(
-            (row['lease_id'], row['start_date'], row['end_date']))
+        gpu_users[user_id].add_lease(**row)
 
-    leases_to_delete = []
-    for user_id in user_gpu_leases:
-        for node_id in user_gpu_leases[user_id]:
-            leases = list(set(user_gpu_leases[user_id][node_id]))
-            leases = list(sorted(leases, key=lambda x: x[1]))
-            stacked_leases = find_stacked_leases(leases)
+    # Check for lease Stacking
```

Comment:
This next 20 lines of code or so can be simplified a bit by changing how much state you're managing. It works fine as you have it, just showing an alternative here:

```python
lease_delete_count = 0
for gpu_user in gpu_users.values():
    leases_in_violation = gpu_user.leases_in_violation()
    if leases_in_violation:
      if not describe:
        lease_delete_count += len(leases_in_violation)
        lease_delete(leases_in_violation)
        send_delete_notification(gpu_user, leases=leases_in_violation, sender=sender)
      else:
        pprint(gpu_user.print_info())
```
    
The main change is having the GPU user model report the leases that it knows are in violation. Note that in this way, the sending of email and deleting of leases takes place out of the domain of the GPU user model, as part of a private helper function or something. The GPU user just knows what constitutes lease stacking on a given user and returns that output for something else to do with as they please. That's how I would do it, keeping the user model pretty stupid and devoid of business logic. For something as self-contained as a hammer, these distinctions are not as important, but I thought I would at least point it out as this is one of the objectives of code review :)

Reply:
This sounds good. I'll implement and merge